src/meili/routing.cc

```diff
@@ -524,43 +524,91 @@ find_shortest_path(baldr::GraphReader& reader,
+          const baldr::GraphTile* start_tile = nullptr;
+          const auto directed_edge = reader.directededge(origin_edge.id, start_tile);
+
+          baldr::GraphId opposite_edgeid = reader.GetOpposingEdgeId(origin_edge.id);
+          const baldr::GraphTile* end_tile = directed_edge->leaves_tile()
+                                                 ? reader.GetGraphTile(directed_edge->endnode())
+                                                 : start_tile;
+          const baldr::DirectedEdge* opposite_edge = reader.directededge(opposite_edgeid, end_tile);
 
           // Skip if edge is not allowed
-          if (!directededge || !IsEdgeAllowed(directededge, origin_edge.id, costing, label, tile)) {
+          if (!directed_edge ||
+              !IsEdgeAllowed(directed_edge, origin_edge.id, costing, label, start_tile)) {
             continue;
           }
 
           // U-turn cost
           float turn_cost = label.turn_cost();
           if (label.edgeid().Is_Valid() && label.edgeid() != origin_edge.id &&
-              label.opp_local_idx() == directededge->localedgeidx()) {
+              label.opp_local_idx() == directed_edge->localedgeidx()) {
             turn_cost += turn_cost_table[0];
           }
 
           // All destinations on this origin edge
           for (const auto other_dest : edge_dests[origin_edge.id]) {
             // All edges of this destination
-            for (const auto& other_edge : destinations[other_dest].edges) {
-              if (origin_edge.id == other_edge.id &&
-                  origin_edge.percent_along <= other_edge.percent_along) {
+            for (const auto& destination_edge : destinations[other_dest].edges) {
+              // when expanding origin edges, there are two possible scenarios:
+              // 1. destination_edge is on the same edge with the origin edge, this indicates that
+              //    the path finding algorithm may find origin destination match on same edge if
+              //    percentage flow agrees topologically (see below):
+              //
+              //    origin percent_along
+              //             |
+              //  0% X------------------------------------------------>X 100%  (directed_edge)
+              //                                  |
+              //                          destination percent_along
+              //
+              // 2. destination_edge is on the opposited edge of the origin edge, this indicates that
+              //    an u-turn may occur if percentage flow agrees topologically (see below):
+              //
+              //                            origin percent_along
+              //                                      |
+              // 0%   X------------------------------------------------>X 100% (directed_edge)
+              // 100% X<------------------------------------------------X 0%   (opposited_edge)
+              //               |
+              //    destination percent_along
+
```

Comment:
This is the u-turn under flow fix.  label expansion for u turn check

Reply:
Thanks for the Ascii art, it really helps a lot!